packages/inject/src/inject.ts

```diff
@@ -0,0 +1,213 @@
+  eventListener: {
+    handler: (event: WebchatEvent) => void
+    topics: WebchatEventHandlerTopics
+  }
+}
+
+type WebchatEventHandler = (evnt: WebchatEvent) => void
+type WebchatEventHandlerTopics = WebchatEventType[] | ['*']
+
+// full backward compatibility
+const DEFAULT_CHAT_ID = 'bp-web-widget'
+const DEFAULT_IFRAME_ID = 'bp-widget'
+
+const CHAT_REFS: { [chatId: string]: WebchatRef } = {}
+
+function _getContainerId(chatId?: string) {
+  return chatId ? `${chatId}-container` : DEFAULT_CHAT_ID
+}
+
+function _getIframeId(chatId: string) {
+  return chatId || DEFAULT_IFRAME_ID
+}
+
+function _injectDOMElement(
+  tagName: string,
+  selector: string,
+  options: { [key: string]: string } = {}
+): HTMLElement | void {
+  const element = document.createElement(tagName)
+  // @ts-ignore
+  Object.entries(options).forEach(([attrName, attrValue]) => (element[attrName] = attrValue))
+  const parent = document.querySelector(selector)
+  if (!parent) {
+    console.error(`No element correspond to ${selector}`)
+    return
+  }
+  parent.appendChild(element)
+  return element
+}
+
+function _generateIFrameHTML(host: string, config: Config) {
+  const keyStorage = `bp-chat-key-${config.clientId}`
+  let encryptionKey = localStorage.getItem(keyStorage)
+
+  if (!encryptionKey) {
+    encryptionKey = _generateRandomString(32)
+    localStorage.setItem(keyStorage, encryptionKey)
+  }
+
+  const options = encodeURIComponent(JSON.stringify({ config: { ...config, encryptionKey } }))
+  const title = encodeURIComponent(config.botConversationDescription || config.botName || 'Chatbot')
+  const iframeSrc = host + '/index.html?options=' + options
+  const iframeId = _getIframeId(config.chatId)
+  return `<iframe id="${iframeId}" title="${title}" frameborder="0" src="${iframeSrc}" class="bp-widget-web"/>`
+}
+
+function _generateRandomString(length: number) {
+  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789'
+
+  let str = ''
+  for (let i = 0; i < length; i++) {
+    str += chars.charAt(Math.floor(Math.random() * chars.length))
+  }
+
+  return str
+}
+
+function _makeChatRefProxy(chatId: string, target: Partial<WebchatRef>): WebchatRef {
+  const chatRefHandler: ProxyHandler<WebchatRef> = {
+    get(target, prop: keyof WebchatRef) {
+      if (target[prop]) {
+        return target[prop]
+      }
+
+      if (prop === 'postMessage') {
+        return () => {
+          console.warn(
+            `No webchat with id ${chatId} has been initialized. \n Please use window.botpressWebChat.init first.`
+          )
+        }
+      } else if (prop === 'eventListener') {
+        return {
+          handler: (evnt: WebchatEvent) => console.log('this is the default event handler', evnt),
+          types: []
+        }
+      }
+    },
+    set(target, prop: keyof WebchatRef, value) {
+      target[prop] = value
+      return true
+    }
+  }
+  return new Proxy<Partial<WebchatRef>>(target, chatRefHandler) as WebchatRef
```

Comment:
Considering that you return a partial `WebchatRef` why cast it into a `WebchatRef`?

Reply:
I don't return a partial, I return a proxy that behaves like a full WebchatRef